src/validation/rules/KnownDirectives.js

```diff
@@ -31,29 +34,42 @@ export function misplacedDirectiveMessage(
  * A GraphQL document is only valid if all `@directives` are known by the
  * schema and legally positioned.
  */
-export function KnownDirectives(context: ValidationContext): ASTVisitor {
+export function KnownDirectives(
+  context: ValidationContext | SDLValidationContext,
+): ASTVisitor {
+  const locationsMap = Object.create(null);
+  const schema = context.getSchema();
+  const definedDirectives = schema
+    ? schema.getDirectives()
+    : specifiedDirectives;
+  for (const directive of definedDirectives) {
+    locationsMap[directive.name] = directive.locations;
+  }
+
   return {
+    Document(node) {
+      for (const def of node.definitions) {
+        if (def.kind === Kind.DIRECTIVE_DEFINITION) {
+          locationsMap[def.name.value] = def.locations.map(name => name.value);
+        }
+      }
+    },
     Directive(node, key, parent, path, ancestors) {
-      const directiveDef = find(
-        context.getSchema().getDirectives(),
-        def => def.name === node.name.value,
-      );
-      if (!directiveDef) {
+      const name = node.name.value;
+      const locations = locationsMap[name];
```

Comment:
@mjmahone It's unnecessary complication since you `context` has `getDocument` function. So I just removed `Document` and move this code on the top level.